src/off_chain/cow_endpoint_surplus.py

```diff
@@ -122,94 +105,82 @@ def get_solver_competition_data(
 
         return solver_competition_data
 
-    def get_order_data(
-        self, individual_win_order_id: str
-    ) -> Tuple[Dict[str, Any], int]:
+    def get_decoded_settlement(self, tx_hash: str):
         """
-        This function uses an orderUID to fetch order data from CoW endpoint.
-        Checks both production and staging.
+        Takes settlement hash as input, returns decoded settlement data.
         """
+        encoded_transaction = self.web_3.eth.get_transaction(tx_hash)
+        decoded_settlement = DecodedSettlement.new(
+            self.contract_instance, encoded_transaction.input
+        )
+        return (
+            decoded_settlement.trades,
+            decoded_settlement.clearing_prices,
+            decoded_settlement.tokens,
+        )
 
-        individual_order_data = {}
-        status_code = 0
-        try:
-            prod_order_data_url = (
-                f"https://api.cow.fi/mainnet/api/v1/orders/{individual_win_order_id}"
-            )
-            prod_order = requests.get(
-                prod_order_data_url,
-                headers=header,
-                timeout=30,
-            )
-            if prod_order.status_code == 200:
-                individual_order_data = json.loads(prod_order.text)
-                status_code = prod_order.status_code
-
-            elif prod_order.status_code == 404:
-                barn_order_data_url = (
-                    "https://barn.api.cow.fi/mainnet/api/v1"
-                    f"/orders/{individual_win_order_id}"
-                )
-                barn_order = requests.get(
-                    barn_order_data_url, headers=header, timeout=30
-                )
-                if barn_order.status_code == 200:
-                    individual_order_data = json.loads(barn_order.text)
-                    status_code = barn_order.status_code
-        except ValueError as except_err:
-            self.logger.error(
-                "Endpoint might be down, Unhandled exception: %s.", str(except_err)
-            )
-
-        return (individual_order_data, status_code)
+    def get_onchain_order_data(self, trade, onchain_clearing_prices, tokens):
+        """
+        Returns required data to calculate surplus for winning order
+        using onchain data from decoded settlement.
+        """
+        return {
+            "sell_token_clearing_price": onchain_clearing_prices[
+                trade["sellTokenIndex"]
+            ],
+            "buy_token_clearing_price": onchain_clearing_prices[trade["buyTokenIndex"]],
+            "sell_token": tokens[trade["sellTokenIndex"]].lower(),
+            "buy_token": tokens[trade["buyTokenIndex"]].lower(),
+            "order_type": str(f"{trade['flags']:08b}")[-1],
+            "executed_amount": trade["executedAmount"],
+            "sell_amount": trade["sellAmount"],
+            "buy_amount": trade["buyAmount"],
+            "fee_amount": trade["feeAmount"],
+        }
```

Comment:
When using trades from the decoded settlement, `trades` has to be accessed with integers. I deduced the numbers from `gpv2_settlement.py`. The tests pass now.
```suggestion
        return {
            "sell_token_clearing_price": onchain_clearing_prices[
                trade[0]
            ],
            "buy_token_clearing_price": onchain_clearing_prices[trade[1]],
            "sell_token": tokens[trade[0]].lower(),
            "buy_token": tokens[trade[1]].lower(),
            "order_type": str(f"{trade[8]:08b}")[-1],
            "executed_amount": trade[9],
            "sell_amount": trade[3],
            "buy_amount": trade[4],
            "fee_amount": trade[7],
        }
```

Reply:
which tests are you referring to? (It would work either way though, via integer, or field)

Reply:
When I run `python3 -m tests.unit.test` on the code of this pull request I get the error
```
File "/home/felix/Work/Code/ebbo/src/off_chain/cow_endpoint_surplus.py", line 134, in get_onchain_order_data
    trade["sellTokenIndex"]
TypeError: tuple indices must be integers or slices, not str
```
If I index with integers instead, I get no error.
Can this be due to having a different version of dependencies installed? If the test runs for you without errors it is probably that.
In that case we should change the `README.md` to include instuctions on how to install dependencies using `requirements.txt`? Something along the lines of `python3 -m pip install requirements.txt`.

Reply:
Ok, I tried it and the test passes after installing dependencies using the command `python3 -m pip install -r requirements.txt`.

Reply:
Oh interesting. We should sort this out.